api/websocket/messagebuffer/messagebuffer.go

```diff
@@ -2,38 +2,268 @@ package messagebuffer
 
 import (
 	"encoding/hex"
+	"fmt"
+	"io"
 	"sync"
+	"time"
 
+	rbt "github.com/emirpasic/gods/trees/redblacktree"
+	"github.com/nknorg/nkn/v2/config"
 	"github.com/nknorg/nkn/v2/pb"
+	"github.com/nknorg/nkn/v2/util/log"
 )
 
+const maxSeqId int64 = 1000000 // it should be less than 1 million messages per milli-second
+const MonitorInterval = 10     // seconds
+
+// msg node with create time and expire time
+type msgNode struct {
+	msg      *pb.Relay
+	createId int64
+	expireId int64
+}
+
+func (m *msgNode) GetSize() int {
+	// m.createId(int64), m.expireId(int64), m.msg.MaxHoldingSeconds(int32), m.msg.SigChainLen(int32)
+	return (8 + 8 + 4 + 4 + len(m.msg.SrcIdentifier) + len(m.msg.SrcPubkey) + len(m.msg.DestId) +
+		len(m.msg.Payload) + len(m.msg.BlockHash) + len(m.msg.LastHash))
+}
+
+type indexNode struct {
+	clientID string
+	createId int64
+	expireId int64
+}
+
 // MessageBuffer is the buffer to hold message for clients not online
+// here use several locks, must use locks in the same order to avoid dead lock.
+// the order should be: msgMu -> createMu -> expireMu
 type MessageBuffer struct {
 	sync.Mutex
-	buffer map[string][]*pb.Relay
+	buffer     map[string]*rbt.Tree // map ClinetId to msgs tree
+	treeCreate *rbt.Tree            // index tree of create time
+	treeExpire *rbt.Tree            // index tree of expire time
+
+	totalBufSize int   // the buffer size in bytes
```

Comment:
`totalBufSize` and `seqId` are also protected by lock

Reply:
yes. totalBufSize is changed when tree changed. seqId is only used by AddMessage. so they are locked, unlocked in the tree operation. use the same lock.